### PR TITLE
Rollup

### PR DESCRIPTION
**Pull request by jeremydeanlakey**
This is just a really rough initial implementation.  We'll likely want to do a lot of fine tuning.

### CODE REVIEW

File: BlueDBOnDisk/src/main/java/io/bluedb/disk/segment/Segment.java [hunk 1]
```diff
@@ -116,6 +116,47 @@ public T get(BlueKey key) throws BlueDbException {
...
+		List<File> filesToRollup = getFilesInRange(start, end);
+		Path path = Paths.get(segmentPath.toString(), start + "_" + end);
+		Path tmpPath = FileManager.createTempFilePath(path);
+		List<BlueEntity<T>> entities = new ArrayList<>();
+
+		// TODO switch to streaming
+		// TODO keep sorted, here and everywhere you write
+		for (File file: filesToRollup) {
+			List<BlueEntity<T>> es = fetch(file);
+			entities.addAll(es);
```
Comment (d-w-johnson) [inline BlueDBOnDisk/src/main/java/io/bluedb/disk/segment/Segment.java]: I'd rather not have to hold all entities being rolled up in memory at one time. Instead, we could stream through each file one at a time while writing each object to the new temp file. Holding a write lock for awhile on that temp file shouldn't be an issue.

File: BlueDBOnDisk/src/main/java/io/bluedb/disk/segment/Segment.java [hunk 2]
```diff
@@ -116,6 +116,47 @@ public T get(BlueKey key) throws BlueDbException {
...
+
+		List<BlueWriteLock<Path>> sourceFileWriteLocks = new ArrayList<>();
+		try (BlueWriteLock<Path> targetFileLock = lockManager.acquireWriteLock(path)){
+			for (File file: filesToRollup) {
+				sourceFileWriteLocks.add(lockManager.acquireWriteLock(file.toPath()));
+			}
+
+			FileManager.moveFile(tmpPath, targetFileLock);
+			for (BlueWriteLock<Path> writeLock: sourceFileWriteLocks) {
+				FileManager.deleteFile(writeLock);
```
Comment (d-w-johnson) [inline BlueDBOnDisk/src/main/java/io/bluedb/disk/segment/Segment.java]: Could something be added to one of the files being rolled up after we've already added all of its entities to our list? If so, they didn't make it into the temp file and we'll lose them.

File: BlueDBOnDisk/src/main/java/io/bluedb/disk/segment/RollupScheduler.java [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+import java.util.Map;
+import java.util.Map.Entry;
+import java.util.concurrent.ConcurrentHashMap;
+import io.bluedb.api.exceptions.BlueDbException;
+import io.bluedb.disk.collection.BlueCollectionImpl;
+
+public class RollupScheduler {
+
+	private static final long WAIT_BETWEEN_REVIEWS = 3_000; // TODO something more sophisticated than just wait ?
+	private static final long WAIT_BEFORE_ROLLUP = 3_000; // TODO
```
Comment (d-w-johnson) [inline BlueDBOnDisk/src/main/java/io/bluedb/disk/segment/RollupScheduler.java]: This should start much higher right?

File: BlueDBOnDisk/src/main/java/io/bluedb/disk/segment/Segment.java (hunk 1 above)
Comment (jeremydeanlakey) [reply BlueDBOnDisk/src/main/java/io/bluedb/disk/segment/Segment.java]: using streaming in the latest updates

File: BlueDBOnDisk/src/main/java/io/bluedb/disk/segment/Segment.java (hunk 2 above)
Comment (jeremydeanlakey) [reply BlueDBOnDisk/src/main/java/io/bluedb/disk/segment/Segment.java]: not right now.  the rollup is currently done in the write thread.  If we change that, we'll have to change a bunch of logic.

**Pull request merged**